### PR TITLE
BE-41 Implement endpoint for files and text

### PR DESCRIPTION
**Pull request by Scr0ll**

### CODE REVIEW

File: antiplag/.env.example [hunk 1]
```diff
@@ -1,3 +1,3 @@
 DEBUG=True
 DJANGO_SECRET_KEY=3YSxRH/mcqSJpkk40s+0qy+uh+9o5LIBdZtwHOo6oNaLe8+Mx6
-DATABASE_URL=postgresql://postgres:postgres@localhost:5432/antiplag
+DATABASE_URL=postgresql://postgres:qwelkj125@127.0.0.1:5432/antiplag
```
Comment (OliverKanat) [inline antiplag/.env.example:3]: Keep original value

File: antiplag/antiplag/views.py [hunk 2]
```diff
@@ -28,22 +30,45 @@ class FileDetail(APIView):
...
     def post(self, request, *args, **kwargs):
-
-        submission_serializer = SubmissionSerializer(data=request.data)
+        submission_serializer = SubmissionSerializer(data={"status": Submission.SubmissionStatus.PENDING})
         if submission_serializer.is_valid():
             submission = submission_serializer.save()
-            for file in request.FILES.getlist("files"):
-                doc_serializer = DocumentSerializer(data={"file": file})
+
+            if request.META['HTTP_TYPE'] == Document.DocumentType.FILE:
```
Comment (OliverKanat) [inline antiplag/antiplag/views.py:37]: Use https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type, not our interal ENUM

File: antiplag/antiplag/views.py [hunk 3]
```diff
@@ -28,22 +30,45 @@ class FileDetail(APIView):
...
+                for file in request.FILES.getlist("files"):
+                    text = Document.process_file(self, file)
+                    text_raw = Document.process_raw_text(self, text)
+                    language = Document.detect_language(self, text_raw)
```
Comment (OliverKanat) [inline antiplag/antiplag/views.py:41]: Why dont you use `Document.create_and_process_text()`?

File: antiplag/antiplag/views.py [hunk 4]
```diff
@@ -28,22 +30,45 @@ class FileDetail(APIView):
...
+                    doc_serializer = DocumentSerializer(data={"file": file,
+                                                              "type": Document.DocumentType.FILE,
+                                                              "text": text,
+                                                              "text_raw": text_raw,
+                                                              "language": language
+                                                              })
+                    if doc_serializer.is_valid():
+                        document = doc_serializer.save()
+                        document.submission = submission
+                        document.save()
```
Comment (OliverKanat) [inline antiplag/antiplag/views.py:51]: Cant the submission be added before serializer checks validity?

File: antiplag/antiplag/views.py [hunk 5]
```diff
@@ -28,22 +30,45 @@ class FileDetail(APIView):
...
+                return Response(SubmissionSerializer(submission).data, status=status.HTTP_201_CREATED)
+            elif request.META['HTTP_TYPE'] == Document.DocumentType.TEXT:
+
+                text = request.POST.get('text', "no_text_attribute")
+                raw_text = request.POST.get('text', "no_text_attribute")
```
Comment (OliverKanat) [inline antiplag/antiplag/views.py:56]: Get only raw text and parse it using `Document. create_and_process_text()`. If no text is sent, use empty string instead. Or maybe it should just fail?

File: antiplag/antiplag/views.py [hunk 6]
```diff
@@ -28,22 +30,45 @@ class FileDetail(APIView):
...
+                language = Document.detect_language(self, raw_text)
```
Comment (OliverKanat) [inline antiplag/antiplag/views.py:57]: Use it in `Document. create_and_process_text()`.

File: antiplag/antiplag/views.py [hunk 7]
```diff
@@ -28,22 +30,45 @@ class FileDetail(APIView):
...
+                doc_serializer = DocumentSerializer(data={"type": Document.DocumentType.FILE,
+                                                          "text": text,
+                                                          "text_raw": raw_text,
+                                                          "language": language
+                                                          })
                 if doc_serializer.is_valid():
                     document = doc_serializer.save()
                     document.submission = submission
                     document.save()
```
Comment (OliverKanat) [inline antiplag/antiplag/views.py:66]: Same as before, cant the submission be added before validation?

File: antiplag/antiplag/views.py [hunk 8]
```diff
@@ -28,22 +30,45 @@ class FileDetail(APIView):
...
-
-            return Response(
-                SubmissionSerializer(submission).data, status=status.HTTP_201_CREATED
-            )
-        else:
-            return Response(status=status.HTTP_400_BAD_REQUEST)
+                else:
+                    print(doc_serializer.errors)
```
Comment (OliverKanat) [inline antiplag/antiplag/views.py:68]: Else return bad request - code 400.

File: antiplag/antiplag/views.py (hunk 4 above)
Comment (OliverKanat) [inline antiplag/antiplag/views.py:51]: What if validation fails?

File: antiplag/antiplag/models.py [hunk 9]
```diff
@@ -62,3 +63,6 @@ def process_file(self, file):
     def process_raw_text(self, text_raw):
         # Replace with actual process raw text method.
         return text_raw
+
+    def detect_language(self, text_raw):
+        return detect(text_raw)
```
Comment (OliverKanat) [inline antiplag/antiplag/models.py:68]: Use it directly in `Document.Document. create_and_process_text()`.

**Review by OliverKanat — CHANGES_REQUESTED** [review]

**Comment by MCFreddie777** [comment]
@OliverKanat Following line also matches `/file/upload` as well as `/upload`.
Could we make it more strict?

```python
url(r"^upload/$", FileDetail().as_view(), name="upload")
```

File: antiplag/models.py [hunk 10]
```diff
@@ -47,24 +47,39 @@ def create_and_process_text(cls, submission=None, text_raw=None, file=None, lang
...
+            type=type,
+            language=language
+        )
+        document = document.save()
+
         return document
 
     def __str__(self):
-        return f"document-{self.id}-{self.type.label}"
+        return f"document-{self.id}-{self.type}"
```
Comment (OliverKanat) [inline antiplag/models.py:65]: `label` attribute is usually for human readable representation of `TextChoices`.

**Review by OliverKanat — APPROVED** [review]

File: antiplag/models.py (hunk 10 above)
Comment (MCFreddie777) [reply antiplag/models.py:65]: Changed back to `.label` ✅

**Review by OliverKanat — APPROVED** [review]

**Pull request merged**